File: nodedoc/docentry-storage/src/main/java/at/scch/nodedoc/db/documents/TextId.java
package at.scch.nodedoc.db.documents;

import at.scch.nodedoc.nodeset.DefinitionField;
import at.scch.nodedoc.nodeset.UAVariable;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Getter
@Setter
@EqualsAndHashCode
@ToString
public class TextId {
    public static class Type {
        public static final String DOCUMENTATION = "documentation";
        public static final String DESCRIPTION = "description";
        public static final String ARGUMENT = "argument";
        public static final String FIELD = "field";
        public static final String FREETEXT = "freetext";
    }

    private String type;
    private String nodeId;
    private String fieldName;
    private String argumentName;
    private String headerId;

    private TextId(String type, String nodeId, String fieldName, String argumentName, String headerId) {
        this.type = type;
        this.nodeId = nodeId;
        this.fieldName = fieldName;
        this.argumentName = argumentName;
        this.headerId = headerId;
    }

    public static TextId forNodeDescription(at.scch.nodedoc.nodeset.NodeId<?> nodeId) {
        return new TextId(Type.DESCRIPTION, nodeId.toExpandedNodeIdString(), null, null, null);
    }

    public static TextId forNodeDocumentation(at.scch.nodedoc.nodeset.NodeId<?> nodeId) {
        return new TextId(Type.DOCUMENTATION, nodeId.toExpandedNodeIdString(), null, null, null);
    }

    public static TextId forField(at.scch.nodedoc.nodeset.NodeId<?> dataTypeNodeId, DefinitionField field) {
        return new TextId(Type.FIELD, dataTypeNodeId.toExpandedNodeIdString(), field.getName(), null, null);
    }

    public static TextId forArgument(at.scch.nodedoc.nodeset.NodeId<?> methodNodeId, UAVariable.Argument argument) {
        return new TextId(Type.ARGUMENT, methodNodeId.toExpandedNodeIdString(), null, argument.getName(), null);
    }

    public static TextId forFreetext(String headerId) {
        return new TextId(Type.FREETEXT, null, null, null, headerId);
    }
}

File: nodedoc/documentation/src/main/docscript/doc-entry-editor/TextId.ts
type DocumentationTextId = {
    type: "documentation",
    nodeId: string,
}

type DescriptionTextId = {
    type: "description",
    nodeId: string,
}

type FieldTextId = {
    type: "field",
    nodeId: string,
    fieldName: string,
}

type ArgumentTextId = {
    type: "argument",
    nodeId: string,
    argumentName: string,
}

type FreetextTextId = {
    type: "freetext",
    headerId: string,
}

export type TextId = 
    DocumentationTextId | 
    DescriptionTextId | 
    ArgumentTextId | 
    FreetextTextId | 
    FieldTextId;

export function isTextIdEqual(a: TextId, b: TextId) {
    return (a.type === "documentation" && b.type === "documentation" && a.nodeId === b.nodeId)
        || (a.type === "description" && b.type === "description" && a.nodeId === b.nodeId)
        || (a.type === "field" && b.type === "field" && a.nodeId === b.nodeId && a.fieldName === b.fieldName)
        || (a.type === "argument" && b.type === "argument" && a.nodeId === b.nodeId && a.argumentName === b.argumentName)
        || (a.type === "freetext" && b.type === "freetext" && a.headerId === b.headerId)
}